Add verify check for inconsistent settings

Hiviexd opened the pull request:
Part of https://github.com/ppy/osu/issues/12091

Ports [the following check](https://github.com/Naxesss/MapsetVerifier/blob/6e48f1269b0a484490a0ddc75df23b3400037a3d/src/Checks/AllModes/Settings/CheckInconsistentSettings.cs) from MV.

Initially thought about integrating it with the main metadata check but felt it'd be better to keep concerns separate.

Doesn't include every setting in the MV check since they're either:
- Not easily checkable AND they're too rare to be worth the effort
- No longer a concern in lazer (namely the beatmap ID check)

Hiviexd committed:
add check for inconsistent settings

Hiviexd committed:
add tests

Hiviexd committed:
fix tests

bdach commented on osu.Game/Rulesets/Edit/Checks/CheckInconsistentSettings.cs:
```diff
@@ -0,0 +1,80 @@
+            // Define fields to check
+            var fieldsToCheck = new (IssueType issueType, string fieldName, Func<IBeatmap, string> fieldSelector)[]
+            {
+                (IssueType.Warning, "Audio lead-in", b => b.AudioLeadIn.ToString(CultureInfo.InvariantCulture)),
+                (IssueType.Warning, "Countdown", b => b.Countdown.ToString()),
+                (IssueType.Warning, "Countdown offset", b => b.CountdownOffset.ToString()),
+                (IssueType.Warning, "Epilepsy warning", b => b.EpilepsyWarning.ToString()),
+                (IssueType.Warning, "Letterbox during breaks", b => b.LetterboxInBreaks.ToString()),
+                (IssueType.Warning, "Samples match playback rate", b => b.SamplesMatchPlaybackRate.ToString()),
+                (IssueType.Warning, "Widescreen support", b => b.WidescreenStoryboard.ToString()),
+                (IssueType.Negligible, "Tick Rate", b => b.Difficulty.SliderTickRate.ToString(CultureInfo.InvariantCulture)),
```
going through `ToString()` to check equality of *anything* is criminal, I cannot simply ignore this.

please refactor such that this operates on actual pre-stringed values. will likely involve something like rearranging the loops below so that the inner loop is across difficulties, and some degree of generics usage. probably something along the lines of

```csharp
void checkIssue<T>(IssueType issueType, string fieldName, Func<IBeatmap, T> fieldSelector)
    where T : IEquatable<T>
{
    // ...
}
```

or similar

bdach replied:
Better but not quite. `EqualityComparer<object>.Default` is generally not advised for usage.

Please check https://github.com/ppy/osu/pull/34540/commits/c2a85032a6e06785a43f74624a384410141ed7de and whether it behaves how you'd expect it to. It passes tests so I'd like to think so but I'd rather you check it anyway.

Hiviexd committed:
refactor comparison and use actual object type for field selectors

Hiviexd committed:
move storyboard presence check to shared utils

Hiviexd committed:
only check widescreen bool when storyboard is present

matches MV

Hiviexd committed:
update widescreen support tests

bdach committed:
Use strongly-typed equality comparer rather than using object equality comparer

`EqualityComparer<object>.Default()` will be slow, will fall back to
`object.Equals()` which may do the right thing, the wrong thing, or be
useless due to using reference equality semantics.

In practice in this call site it likely doesn't matter anyway but I'd
rather be future-proof than not.

bdach committed:
Merge branch 'master' into verify/check-inconsistent-settings
